ReproducibleEnvs.Rmd

```diff
@@ -0,0 +1,564 @@
+library(gradethis)
+library(renv)
+knitr::opts_chunk$set(echo = FALSE)
+
+tutorial_options(
+  exercise.checker = gradethis::grade_learnr
+)
+
+```
+
+```{r phs-logo, echo=FALSE, fig.align='right', out.width="40%"}
+knitr::include_graphics("images/phs-logo.png")
+```
+
+## Introduction
+
+Welcome to Reproducible Environments in R. If you've ever had code stop working at the next iteration of a run or it runs fine for you but not a colleague, this course is for you! This course is designed as a self-led dive into reproducible environments: what they are and how to use them, specifically using the `{renv}` package. Throughout this course there will be quizzes to test your knowledge and opportunities to modify and write R code.
+
+::: info_box
+<h4>Course Info</h4>
+
+<ul>
+
+<li>This course is built to flow through sections and build on previous knowledge. If you're comfortable with a particular section, you can skip it.</li>
+
+<li>Most sections have multiple parts to them. Navigate the course by using the buttons at the bottom of the screen to Continue or go to the Next Topic.</li>
+
+<li>The course will also show progress through sections, a green tick will appear on sections you've completed, and it will remember your place if you decide to close your browser and come back later.</li>
+
+</ul>
+:::
+
+</br>
+
+### Learning Outcomes
+
+-   Reinforced understanding of the importance of reproducibility in data projects.
+-   Be able to set up and configure `{renv}` for managing project dependencies.
+-   Create reproducible environments for R projects using `{renv}`.
+-   Understand best practices for version control and collaboration with `{renv}`.
+-   Troubleshoot common issues related to reproducible environments.
+-   Awareness of advanced features for customisation options.
+-   Be able to confidently apply reproducibility principles to improve reliability and transparency of projects.
+
+### What is `{renv}`?
+
+R manages packages through a central library and shares this across all projects. In PHS Posit Workbench, we expand on this slightly, with a system-level library (used for base packages) and then user-level where each user-installed package will go. On any system, this location can be found using the `.libPaths()` function, where the list of locations is listed in order of priority.
+
+This works well until versions of packages are considered. Each project may require different versions of packages, resulting in deprecated or missing functionality. This issue is only made worse the moment the project is shared with someone else, someone with their very own library of packages with different versions. The `{renv}` package solves this issue by creating a separate, reproducible environment that is attached to the project, creating:
+
+-   a project-specific library, and
+-   a list of dependencies and their version.
+
+```{r renv-libraries, echo=FALSE, fig.align='center'}
+knitr::include_graphics("images/renv-libraries.png")
+```
```

Comment:
I found this graphic a bit confusing, I think probably some form of nesting would convey the concept better... 

I had a quick go at a nested hex (fun little diversion), see code and images below. If you agree with my point feel free to use the images/code or trash them and make better ones!

<details>
  <summary>Code to produce</summary>

```r

library(ggplot2)
library(phsstyles)
# Data frame for a hexagon ----
angle <- seq(0, 2 * pi, length.out = 7) + pi / 6
polygon_df <- data.frame(
  angle = angle,
  x     = cos(angle),
  y     = sin(angle)
)

# Colours ----
hex_col_1 <- phs_colors("phs-purple")
hex_col_2 <- phs_colors("phs-magenta")
hex_col_3 <- phs_colors("phs-blue")

transparent_white <- "#FFFFFF00"

# Plots with nested hexagons, sequentially dashed.
p_1 <- ggplot(polygon_df) +
  geom_polygon(
    aes(x = x, y = y),
    linetype = "dashed",
    fill = transparent_white,
    colour = hex_col_1
  ) +
  geom_polygon(
    aes(x = x * 2 / 3, y = y * 2 / 3),
    fill = transparent_white,
    colour = hex_col_2
  ) +
  geom_polygon(
    aes(x = x * 1 / 3, y = y * 1 / 3),
    fill = transparent_white,
    colour = hex_col_3
  ) +
  coord_equal() +
  theme_void()

p_2 <- ggplot(polygon_df) +
  geom_polygon(
    aes(x = x, y = y),
    fill = transparent_white,
    colour = hex_col_1
  ) +
  geom_polygon(
    aes(x = x * 2 / 3, y = y * 2 / 3),
    linetype = "dashed",
    fill = transparent_white,
    colour = hex_col_2
  ) +
  geom_polygon(
    aes(x = x * 1 / 3, y = y * 1 / 3),
    fill = transparent_white,
    colour = hex_col_3
  ) +
  coord_equal() +
  theme_void()

p_3 <- ggplot(polygon_df) +
  geom_polygon(
    aes(x = x, y = y),
    fill = transparent_white,
    colour = hex_col_1
  ) +
  geom_polygon(
    aes(x = x * 2 / 3, y = y * 2 / 3),
    fill = transparent_white,
    colour = hex_col_2
  ) +
  geom_polygon(
    aes(x = x * 1 / 3, y = y * 1 / 3),
    linetype = "dashed",
    fill = transparent_white,
    colour = hex_col_3
  ) +
  coord_equal() +
  theme_void()

p_1
p_2
p_3

# Save plots ----
ggsave("outer_hex.png", p_1)
ggsave("middle_hex.png", p_2)
ggsave("inner_hex.png", p_3)
```
</details>

![inner_hex](https://github.com/Public-Health-Scotland/learnr-online/assets/5982260/423bffbd-c6e5-42af-834d-a409cb8dcf33)
![middle_hex](https://github.com/Public-Health-Scotland/learnr-online/assets/5982260/cdf36de1-f2b9-4c8b-a79b-72b5073a2f6a)
![outer_hex](https://github.com/Public-Health-Scotland/learnr-online/assets/5982260/804b7834-df20-4ce0-bf11-34a91bf9b357)

Reply:
Hey, @Moohan - thanks for your review, I've implemented changes based on your suggestions. 
The current graphic isn't great, I created it to have something visual but I don't like it either so I'm keen to explore alternatives. While I do like the style of your suggestions, I'm hesitant to show the model as a nested one. I understand where you're coming from but I'm concerned it's misleading from how libraries actually work. I'll have a think...